lib/std/parker.zig

```diff
@@ -0,0 +1,381 @@
+                const key = @ptrCast(*const c_void, ptr);
+                var waiting = @atomicLoad(u32, waiters, .Acquire);
+                while (waiting != 0) {
+                    waiting = @cmpxchgWeak(u32, waiters, waiting, waiting - 1, .Acquire, .Monotonic) orelse {
+                        const rc = windows.ntdll.NtReleaseKeyedEvent(self.handle, key, windows.FALSE, null);
+                        assert(rc == 0);
+                        return;
+                    };
+                }
+            }
+
+            fn park(self: KeyedEvent, ptr: *const u32, expected: u32, waiters: *u32) void {
+                const key = @ptrCast(*const c_void, ptr);
+                while (@atomicLoad(u32, ptr, .Acquire) == expected) {
+                    _ = @atomicRmw(u32, waiters, .Add, 1, .Release);
+                    const rc = windows.ntdll.NtWaitForKeyedEvent(self.handle, key, windows.FALSE, null);
+                    assert(rc == 0);
+                }
+            }
+        };
+    };
+};
+
+const PosixParker = struct {
+    cond: pthread_cond_t,
+    mutex: pthread_mutex_t,
+
+    pub fn init() PosixParker {
+        return PosixParker{
+            .cond = PTHREAD_COND_INITIALIZER,
+            .mutex = PTHREAD_MUTEX_INITIALIZER,
+        };
+    }
+
+    pub fn deinit(self: *PosixParker) void {
+        // On dragonfly, the destroy functions return EINVAL if they were initialized statically.
+        const retm = pthread_mutex_destroy(&self.mutex);
+        assert(retm == 0 or retm == (if (builtin.os == .dragonfly) os.EINVAL else 0));
+        const retc = pthread_cond_destroy(&self.cond);
+        assert(retc == 0 or retc == (if (builtin.os == .dragonfly) os.EINVAL else 0));
+    }
+
+    pub fn unpark(self: *PosixParker, ptr: *const u32) void {
+        assert(pthread_mutex_lock(&self.mutex) == 0);
+        defer assert(pthread_mutex_unlock(&self.mutex) == 0);
+        assert(pthread_cond_signal(&self.cond) == 0);
+    }
+
+    pub fn park(self: *PosixParker, ptr: *const u32, expected: u32) void {
+        assert(pthread_mutex_lock(&self.mutex) == 0);
+        defer assert(pthread_mutex_unlock(&self.mutex) == 0);
+        while (@atomicLoad(u32, ptr, .Acquire) == expected)
+            assert(pthread_cond_wait(&self.cond, &self.mutex) == 0);
+    }
+
+    const PTHREAD_MUTEX_INITIALIZER = pthread_mutex_t{};
+    extern "c" fn pthread_mutex_lock(mutex: *pthread_mutex_t) c_int;
+    extern "c" fn pthread_mutex_unlock(mutex: *pthread_mutex_t) c_int;
+    extern "c" fn pthread_mutex_destroy(mutex: *pthread_mutex_t) c_int;
+
+    const PTHREAD_COND_INITIALIZER = pthread_cond_t{};
+    extern "c" fn pthread_cond_wait(noalias cond: *pthread_cond_t, noalias mutex: *pthread_mutex_t) c_int;
+    extern "c" fn pthread_cond_signal(cond: *pthread_cond_t) c_int;
+    extern "c" fn pthread_cond_destroy(cond: *pthread_cond_t) c_int;
+
+    // https://github.com/rust-lang/libc
+    usingnamespace switch (builtin.os) {
```

Comment:
If you could do it, that would be better since am not sure what to do with the definitions for the other platforms like solaris and fuchsia (drop them for now since they're low tier support?)

Reply:
OK no problem, I'll do it as a final fixup when merging. Let me know when it's ready and I'll have another look. Since you went through the trouble to provide these definitions, I'll add new subdirectories for these operating systems. Thanks to Zig's lazy analysis, this will not cause any issues; we can slowly build up support for these other systems over time.